File: examples/CSharpProd/HttpTests/AdvancedHttpTest.cs
using System;
using System.Linq;
using System.Net.Http;
using Elasticsearch.Net.Specification.NodesApi;
using NBomber;
using NBomber.Contracts;
using NBomber.CSharp;
using NBomber.Plugins.Http.CSharp;
using NBomber.Plugins.Network.Ping;
using Newtonsoft.Json;

namespace CSharpProd.HttpTests
{
    // in this example we use:
    // - NBomber.Http (https://nbomber.com/docs/plugins-http)

    public class UserResponse
    {
        public string Id { get; set; }
        public string Name { get; set; }
        public string Email { get; set; }
        public string Phone { get; set; }
    }

    public class PostResponse
    {
        public string Id { get; set; }
        public string UserId { get; set; }
        public string Title { get; set; }
        public string Body { get; set; }
    }

    public class AdvancedHttpTest
    {
        public static void Run()
        {
            var userFeed = Feed.CreateRandom(
                name: "userFeed",
                new[] {"1", "2", "3", "4", "5"}
            );

            var httpFactory = HttpClientFactory.Create();

            var getUser = Step.Create("get_user", clientFactory: httpFactory, userFeed, async context =>
            {
                var userId = context.FeedItem;
                var url = $"https://jsonplaceholder.typicode.com/users?id={userId}";

                var request = Http.CreateRequest("GET", url)
                    .WithCheck(async response =>
                    {
                        var json = await response.Content.ReadAsStringAsync();

                        // parse JSON
                        var users = JsonConvert.DeserializeObject<UserResponse[]>(json);


                        return users?.Length == 1
                            ? Response.Ok(users.First()) // we pass user object response to the next step
                            : Response.Fail($"not found user: {userId}");
                    });
                var response = await Http.Send(request, context);
                return response;


            });

            // this 'getPosts' will be executed only if 'getUser' finished OK.
            var getPosts = Step.Create("get_posts", clientFactory: httpFactory, async context =>
            {
                var user = context.GetPreviousStepResponse<UserResponse>();
                var url = $"https://jsonplaceholder.typicode.com/posts?userId={user.Id}";

                var request = Http.CreateRequest("GET", url)
                    .WithCheck(async response =>
                    {
                        var json = await response.Content.ReadAsStringAsync();

                        // parse JSON
                        var posts = JsonConvert.DeserializeObject<PostResponse[]>(json);

                        return posts?.Length > 0
                            ? Response.Ok()
                            : Response.Fail($"not found posts for user: {user.Id}");
                    });
                var response = await Http.Send(request, context);
                return response;
            });

            var scenario = ScenarioBuilder
                .CreateScenario("rest_api", getUser, getPosts)
                .WithWarmUpDuration(TimeSpan.FromSeconds(5))
                .WithLoadSimulations(new[]
                {
                    Simulation.InjectPerSec(rate: 100, during: TimeSpan.FromSeconds(30))
                });

            var pingPluginConfig = PingPluginConfig.CreateDefault(new[] {"jsonplaceholder.typicode.com"});
            var pingPlugin = new PingPlugin(pingPluginConfig);

            NBomberRunner
                .RegisterScenarios(scenario)
                .WithWorkerPlugins(pingPlugin)
                .WithTestSuite("http")
                .WithTestName("advanced_test")
                .Run();
        }
    }
}

File: examples/FSharpProd/HttpTests/AdvancedHttpTest.fs
module FSharpProd.HttpTests.AdvancedHttpTest

open FSharp.Control.Tasks.NonAffine
open Newtonsoft.Json

open NBomber
open NBomber.Contracts
open NBomber.FSharp
open NBomber.Plugins.Http.FSharp
open NBomber.Plugins.Network.Ping

// in this example we use:
// - NBomber.Http (https://nbomber.com/docs/plugins-http)

[<CLIMutable>]
type UserResponse = {
    Id: string
    Name: string
    Email: string
    Phone: string
}

[<CLIMutable>]
type PostResponse = {
    Id: string
    UserId: string
    Title: string
    Body: string
}

let run () =

    let userFeed = ["1"; "2"; "3"; "4"; "5"]
                   |> Feed.createRandom "userFeed"

    let httpFactory = HttpClientFactory.create()

    let getUser = Step.create("get_user",
                              clientFactory = httpFactory,
                              feed = userFeed,
                              execute = fun context ->

        let userId = context.FeedItem
        let url = $"https://jsonplaceholder.typicode.com/users?id={userId}"

        Http.createRequest "GET" url
        |> Http.withCheck(fun response -> task {
            let! json = response.Content.ReadAsStringAsync()

            // parse JSON
            let users = json
                        |> JsonConvert.DeserializeObject<UserResponse[]>
                        |> ValueOption.ofObj

            match users with
            | ValueSome usr when usr.Length = 1 ->
                return Response.ok(usr[0]) // we pass user object response to the next step

            | _ ->
                return Response.fail($"not found user: {userId}")
        })
        |> Http.send context
    )

    // this 'getPosts' will be executed only if 'getUser' finished OK.
    let getPosts = Step.create("get_posts",
                               clientFactory = httpFactory,
                               execute = fun context ->

        let user = context.GetPreviousStepResponse<UserResponse>()
        let url = $"https://jsonplaceholder.typicode.com/posts?userId={user.Id}"

        Http.createRequest "GET" url
        |> Http.withCheck(fun response -> task {
            let! json = response.Content.ReadAsStringAsync()

            // parse JSON
            let posts = json
                        |> JsonConvert.DeserializeObject<PostResponse[]>
                        |> ValueOption.ofObj

            match posts with
            | ValueSome ps when ps.Length > 0 -> return Response.ok()
            | _                               -> return Response.fail()
        })
        |> Http.send context
    )

    let pingPluginConfig = PingPluginConfig.CreateDefault ["jsonplaceholder.typicode.com"]
    let pingPlugin = new PingPlugin(pingPluginConfig)

    Scenario.create "rest_api" [getUser; getPosts]
    |> Scenario.withWarmUpDuration(seconds 5)
    |> Scenario.withLoadSimulations [InjectPerSec(rate = 100, during = seconds 30)]
    |> NBomberRunner.registerScenario
    |> NBomberRunner.withWorkerPlugins [pingPlugin]
    |> NBomberRunner.withTestSuite "http"
    |> NBomberRunner.withTestName "advanced_test"
    |> NBomberRunner.run
    |> ignore
